File: Design_Pattern_and_Principles/StrategyPatternExample/ConcreteStrategies.java
class CreditCardPayment implements PaymentStrategy {
    private String cardNumber;

    public CreditCardPayment(String cardNumber) {
        this.cardNumber = cardNumber;
    }

    public void pay(double amount) {
        System.out.println("Processed checkout balance of $" + amount + " using Credit Card ending in " + cardNumber.substring(cardNumber.length() - 4));
    }
}

class PayPalPayment implements PaymentStrategy {
    private String email;

    public PayPalPayment(String email) {
        this.email = email;
    }

    public void pay(double amount) {
        System.out.println("Processed checkout balance of $" + amount + " using PayPal login: " + email);
    }
}

File: Design_Pattern_and_Principles/StrategyPatternExample/ConcreteStrategies.cs
using System;

class CreditCardPayment : IPaymentStrategy {
    private string cardNumber;

    public CreditCardPayment(string cardNumber) {
        this.cardNumber = cardNumber;
    }

    public void Pay(double amount) {
        Console.WriteLine("Processed checkout balance of $" + amount + " using Credit Card ending in " + cardNumber.Substring(cardNumber.Length - 4));
    }
}

class PayPalPayment : IPaymentStrategy {
    private string email;

    public PayPalPayment(string email) {
        this.email = email;
    }

    public void Pay(double amount) {
        Console.WriteLine("Processed checkout balance of $" + amount + " using PayPal login: " + email);
    }
}